crates/shared/src/price_estimation/quasimodo.rs

```diff
@@ -0,0 +1,191 @@
+                        normalize_priority: Some(if self.native_token == query.buy_token {
+                            1
+                        } else {
+                            0
+                        }),
+                        ..Default::default()
+                    },
+                )
+            })
+            .collect();
+
+        let (sell_amount, buy_amount) = match query.kind {
+            OrderKind::Buy => (U256::max_value(), query.in_amount),
+            OrderKind::Sell => (query.in_amount, U256::one()),
+        };
+
+        let orders = BTreeMap::from([(
+            0,
+            OrderModel {
+                sell_token: query.sell_token,
+                buy_token: query.buy_token,
+                sell_amount,
+                buy_amount,
+                allow_partial_fill: false,
+                is_sell_order: query.kind == OrderKind::Sell,
+                fee: FeeModel {
+                    amount: *GAS_PER_ORDER * gas_cost,
+                    token: self.native_token,
+                },
+                cost: CostModel {
+                    amount: *GAS_PER_ORDER * gas_cost,
+                    token: self.native_token,
+                },
+                is_liquidity_order: false,
+            },
+        )]);
+
+        let token_pair = TokenPair::new(query.sell_token, query.buy_token).unwrap();
+        let mut pairs = self.base_tokens.relevant_pairs([token_pair].into_iter());
+        pairs.insert(token_pair);
+
+        let amms = self
+            .pools
+            .fetch(pairs, Block::Recent)
+            .await?
+            .iter()
+            .map(|pool| AmmModel {
+                parameters: AmmParameters::ConstantProduct(ConstantProductPoolParameters {
+                    reserves: BTreeMap::from([
+                        (pool.tokens.get().0, pool.reserves.0.into()),
+                        (pool.tokens.get().1, pool.reserves.1.into()),
+                    ]),
+                }),
+                fee: BigRational::from((
+                    BigInt::from(*pool.fee.numer()),
+                    BigInt::from(*pool.fee.denom()),
+                )),
+                cost: CostModel {
+                    amount: *GAS_PER_UNISWAP * gas_cost,
+                    token: self.native_token,
+                },
+                mandatory: false,
+            })
+            .enumerate()
+            .collect();
+
+        let settlement = self
+            .api
+            .solve(
+                &BatchAuctionModel {
+                    tokens,
+                    orders,
+                    amms,
+                    metadata: None,
+                },
+                Instant::now() + Duration::from_secs(5),
+            )
+            .await?;
+
+        if settlement.orders.is_empty() {
+            return Err(PriceEstimationError::NoLiquidity);
+        }
+
+        let mut cost = self.extract_cost(&settlement.orders[&0].cost)?;
```

Comment:
We probably need to account for additional settlement overhead. The constants we use are in:
https://github.com/gnosis/gp-v2-services/blob/main/crates/shared/src/price_estimation/gas.rs

Reply:
Which constant should I use? I think `INITIALIZATION_COST + SETTLEMENT`?

Reply:
We probably need:
```
+ INITIALIZATION_COST // Call into contract
+ SETTLEMENT // overhead for entering the `settle()` function
+ 2*ERC20_TRANSFER // transfer in and transfer out
```